File: 2020/Python/Day7.py
import os
import sys


def main():
    inf = os.path.dirname(os.getcwd())
    inf = os.path.join(inf, "input")
    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf += ".txt"
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf = os.path.join(os.path.dirname(inf), "input")
        inf = os.path.join(inf, "Day7.txt")
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print("None of the expected inputs exist!", file=sys.stderr)
        return
    else:
        print(f"Using input file {inf}.")

    inputFile = open(inf, 'r')
    bags = {}
    for line in inputFile.readlines():
        contained = {}
        for bag in line[line.index("tain ") + 5:].split(", "):
            if "no other" in bag:
                break
            contained[bag[2:bag.index(" bag")]] = int(bag[0])

        bags[line[:line.index(" bag")]] = contained

    goldContaining = 0
    for bag in bags.keys():
        if checkBag(bags, bag, "shiny gold"):
            goldContaining += 1

    print(f"Found {goldContaining} bags that contain a \"shiny gold\" bag.")

    print(f"A single \"shiny gold\" bag contains {countBags(bags, 'shiny gold') - 1} bags.")


def checkBag(bags, bag, lookingFor):
    for check in bags[bag].keys():
        if check == lookingFor:
            return True
        elif checkBag(bags, check, lookingFor):
            return True
    return False


def countBags(bags, bag):
    count = 1
    for check in bags[bag].keys():
        count += countBags(bags, check) * bags[bag][check]

    return count


if __name__ == '__main__':
    main()

File: 2020/Java/src/aoc/Day7.java
package aoc;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public abstract class Day7 {

	public static void main(String[] args) throws IOException {
		Map<String, Map<String, Integer>> bags = new HashMap<>();
		for (String line : InputReader.readInputFileLines(7)) {
			Map<String, Integer> contained = new HashMap<>();
			for (String bag : line.substring(line.indexOf("tain ") + 5).split(", ")) {
				if (bag.equals("no other bags."))
					break;
				contained.put(bag.substring(2, bag.indexOf(" bag")), Character.getNumericValue(bag.charAt(0)));
			}

			bags.put(line.substring(0, line.indexOf(" bag")), contained);
		}

		int goldContaining = 0;
		for (String bag : bags.keySet()) {
			if (checkBag(bags, bag, "shiny gold"))
				goldContaining++;
		}

		System.out.println("Found " + goldContaining + " bags that contain a \"shiny gold\" bag.");

		System.out.println("A single \"shiny gold\" bag contains " + (countBags(bags, "shiny gold") - 1) + " bags.");
	}

	/**
	 * Checks whether the given bag contains a bag of a given type. This is done by
	 * recursively iterating through the bags in the given bag.
	 * 
	 * @param bags       the map of what bags contain what types of bags.
	 * @param bag        the bag to check.
	 * @param lookingFor the bag to check for.
	 * @return whether or not the given bag contains the bag to look for.
	 */
	private static boolean checkBag(Map<String, Map<String, Integer>> bags, String bag, String lookingFor) {
		for (String check : bags.get(bag).keySet()) {
			if (check.equals(lookingFor)) {
				return true;
			} else if (checkBag(bags, check, lookingFor)) {
				return true;
			}
		}
		return false;
	}

	/**
	 * Counts the number of bags in the given bag. Done by recursively iterating
	 * through its contents.
	 * 
	 * @param bags the map of what bags contain what other bags.
	 * @param bag  the bag to check.
	 * @return the total bag count. Includes the start bag.
	 */
	private static int countBags(Map<String, Map<String, Integer>> bags, String bag) {
		int count = 1;
		for (String check : bags.get(bag).keySet()) {
			count += countBags(bags, check) * bags.get(bag).get(check);
		}
		return count;
	}
}
